coreblocks/lsu/pma.py

```diff
@@ -0,0 +1,70 @@
+from dataclasses import dataclass
+from functools import reduce
+from operator import or_
+from amaranth import *
+
+from coreblocks.params import *
+from transactron.utils import HasElaborate
+from transactron.core import TModule
+
+
+@dataclass
+class PMARegion:
+    """
+    Data class for physical memory attributes contiguous region of memory. Region of memory
+    includes both start end end address.
+
+    Attributes
+    ----------
+    start : int
+        Defines beginning of region.
+    end : int
+        Defines end of region.
+    attrs : Record
+        Attributes for given region. Record should follow adequate layout from `PMALayouts`.
+    """
+
+    start: int
+    end: int
+    mmio: bool
+
+
+class PMAChecker(Elaboratable):
+    """
+    Implementation of physical memory attributes checker. It may or may not be a part of LSU.
+    `PMAChecker` exposes method `ask` that can be used to request information about given memory
+    address. Result of `ask` method contains bit vector of all defined PMAs.
+
+    Attributes
+    ----------
+    addr : Signal
+        Memory address, for which PMAs are requested.
+    result : Record
+        PMAs for given address.
+    """
+
+    def __init__(self, gen_params: GenParams) -> None:
+        # poor man's interval list
+        self.segments = gen_params.pma
+        self.attr_layout = gen_params.get(PMALayouts).pma_attrs_layout
+        self.result = Record(self.attr_layout)
+        self.addr = Signal(gen_params.isa.xlen)
+
+    def elaborate(self, platform) -> HasElaborate:
+        m = TModule()
+
+        outputs = Array([Record(self.attr_layout) for _ in self.segments])
```

Comment:
This don't have to be `Array`. Please see that in line 65 you index `outputs` with the python integer `i` not with signal. So all iterations will be done on generation time, so there is no need to include selection logic for runtime.
```suggestion
        outputs = [Record(self.attr_layout) for _ in self.segments]
```